### PR TITLE
Support --upgradeable-program in anchor test

### PR DESCRIPTION
**Pull request by nickguo**
please advise on how to best test this change

### CODE REVIEW

File: cli/src/lib.rs
Comment (acheroncrypto) [inline cli/src/lib.rs:3038]: This looks good to me but it would be better if you could add the `--upgradeable-program` argument to the `Validator` struct here https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/config.rs#L1030 and remove the extra parameter `test_upgradeable_program` from the `validator_flags` function. That way we'll both have access to the `--upgradeable-program` argument(via `test_validator` argument) which could be used in other places and also remove the extra argument from the function.

Comment (nickguo) [reply cli/src/lib.rs]: Ahh very cool!  Instead of an `anchor test` CLI option, would we enable the feature via `anchor.toml` like this? 
```
[test.validator]
upgradeable_program = true
```

I've made the code changes.  PTAL again as I initially had made a mistake:  `--upgradeable-program` requires 3 params with the third being the upgrade authority's pubkey.

Comment (acheroncrypto) [reply cli/src/lib.rs]: I think we can keep the `--upgradeable-program`(or `--upgradeable`) flag for the `anchor test` command for ease of use(like how it was in the initial PR) but also follow the convention of the `solana-test-validator` arguments for the `[test.validator]` entries.

For example `--account` argument is stored as `Option<Vec<AccountEntry>>` https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/config.rs#L1030-L1032

Similarly we could store the `--upgradeable-program` as `Option<Vec<UpgradeableProgramEntry>>` to keep the convention the same(instead of storing `Option<bool>`).

File: cli/src/config.rs
Comment (acheroncrypto) [inline cli/src/config.rs:1029]: It would be better if we kept the same convention for the test validator arguments.

```suggestion
    pub upgradeable_program: Option<Vec<UpgradeableProgramEntry>>,
```

File: cli/src/lib.rs
Comment (acheroncrypto) [inline cli/src/lib.rs:3059]: We also deploy programs similarly with genesis configuration here https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/lib.rs#L3056-L3069

Can add a `upgradeable: bool` field to `GenesisEntry` https://github.com/coral-xyz/anchor/blob/d1e32674d50005d04160b6fbac6f91de602a8829/cli/src/config.rs#L947-L952

which would allow us to specify

```toml
[[test.genesis]]
address = "Test111111111111111111111111111111111111111"
program = "test.so"
upgradeable = true
```

File: cli/src/lib.rs
Comment (nickguo) [reply cli/src/lib.rs]: It looks like the "primary" programs to test are stored separately from the `test.validator` and `test.genesis` entries. https://github.com/nickguo/anchor/blob/2d00b489b44ce045cbc39468d38a4cf7e94dd823/cli/src/lib.rs#L3039 

This is making it a bit tricky as the new `Option<Vec<UpgradeableProgramEntry>>` doesn't have an entry for the main program that we want to test.  So for the main program, we'd have to use something like the initial `    test_upgradeable_program: bool` param to `validator_flags()` which wasn't ideal.

Overall it looks a bit of a code smell -- it should be possible to remove `test.genesis` altogether, and have all programs-to-load configured under `test.validator`.  The main programs can populated there by default.  On top of `UpgradeableProgramEntry` would also add `Option<Vec<BpfProgramEntry>>` which has been missing (functionally the same as the current `test.genesis`).

Sorry for the long post!  I see two ways forward:
(1) Configure upgradeable main program differently from "genesis" programs (e.g. pyth).  For upgradeable main program, we would pass a CLI flag to `anchor test`.  For genesis programs, we have that optional bool.

(2) Refactor away `test.genesis` and instead populate `test.validator`.  We can do this in a backwards compatible way.  And then main program and additional genesis programs are loaded the same way.  Finally, we can add the upgradeable functionality afterwards.

What do you think?

Comment (acheroncrypto) [reply cli/src/lib.rs]: > (1) Configure upgradeable main program differently from "genesis" programs (e.g. pyth). For upgradeable main program, we would pass a CLI flag to `anchor test`. For genesis programs, we have that optional bool.

If I understand correctly, option 1 is roughly the same as [my last post](https://github.com/coral-xyz/anchor/pull/2642#discussion_r1341949281) but both options seem viable.

Comment (nickguo) [reply cli/src/lib.rs]: Went with (1) !  

I decided to bubble up the flag from `Validator` into `TestValidator` because the tricky thing was that `Validator` does not have entries for the programs to deploy.  PTAL 🙏

File: cli/src/config.rs
Comment (nickguo) [reply cli/src/config.rs]: After pushing up the `upgradeable_program` setting from `Validator` into `TestValidator` and `TestValidator.genesis`, we no longer need `UpgradeableProgramEntry` nor `BpfProgramEntry`

File: cli/src/lib.rs
Comment (nickguo) [reply cli/src/lib.rs:3054]: done!  added `upgradeable: bool` to `GenesisEntry`

File: cli/src/lib.rs
Comment (nickguo) [reply cli/src/lib.rs]: And went with `anchor.toml` to configure whether to use upgradeable program.  The CLI flag felt clunky in comparison -- both the ergonomics to use it and the code changes to add it.

Updated `docs/src/pages/docs/manifest.md` to add the new setting

Comment (acheroncrypto) [reply cli/src/lib.rs]: Awesome, thank you for updating the docs as well!

File: docs/src/pages/docs/manifest.md
Comment (acheroncrypto) [inline docs/src/pages/docs/manifest.md:125]: This looks incorrect based on the implementation here https://github.com/coral-xyz/anchor/blob/65032e28999637ee0862a448af710eb30ccedf05/cli/src/config.rs#L952-L959

```suggestion
upgradeable = true
```

**Review by acheroncrypto — COMMENTED** [review]
LGTM other than a couple of small issues I've mentioned.

Could you add an entry to the [CHANGELOG](https://github.com/coral-xyz/anchor/blob/master/CHANGELOG.md#features) before we merge?

File: docs/src/pages/docs/manifest.md
Comment (nickguo) [reply docs/src/pages/docs/manifest.md:125]: yes my bad, thank you for catching!  fixed

File: docs/src/pages/docs/manifest.md
Comment (nickguo) [reply docs/src/pages/docs/manifest.md:138]: updated to just `upgradeable` to match `GenesisEntry` 👍

File: CHANGELOG.md
Comment (acheroncrypto) [inline CHANGELOG.md:31]:
```suggestion
- cli: Add `test.upgradeable`, `test.genesis.upgradeable` setting in anchor.toml to support testing upgradeable programs ([#2642](https://github.com/coral-xyz/anchor/pull/2642)).
```

**Review by acheroncrypto — CHANGES_REQUESTED** [review]
There is a conflict in the [CHANGELOG](https://github.com/coral-xyz/anchor/blob/master/CHANGELOG.md) file that blocks the merge. Could you rebase to master?

File: CHANGELOG.md
Comment (nickguo) [reply CHANGELOG.md:32]: fixed!

File: cli/src/config.rs
Comment (acheroncrypto) [inline cli/src/config.rs:958]: This should be an optional field otherwise it breaks the existing `Anchor.toml` files with genesis entries. https://github.com/coral-xyz/anchor/actions/runs/6386297008/job/17343757677?pr=2642#step:10:297

```
Error: TOML parse error at line 1, column 1
  |
1 | [[test.genesis]]
  | ^^^^^^^^^^^^^^^^
missing field `upgradeable`
```

Comment (nickguo) [reply cli/src/config.rs]: ahh my bad again, fixed!

**Review by acheroncrypto — APPROVED** [review]

**Comment by acheroncrypto** [comment]
Thank you for the contribution and addressing the feedback quickly!

**Pull request merged**
